File: src_nextgen/actool_linux/bom.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
import struct


class BOMError(ValueError):
    """Raised when a BOMStore container is malformed or unsupported."""


@dataclass(frozen=True)
class BOMHeader:
    version: int
    block_count_hint: int
    index_offset: int
    index_length: int
    variables_offset: int
    variables_length: int


@dataclass(frozen=True)
class Block:
    identifier: int
    offset: int
    length: int


class BOMStore:
    """Bounds-checked reader for the BOMStore container used by Assets.car.

    BOM scalar metadata is big-endian. The payload of a named block may use a
    different byte order and is intentionally left uninterpreted by this layer.
    """

    MAGIC = b"BOMStore"
    HEADER = struct.Struct(">8s6I")

    # Known CoreUI database names
    DATABASE_NAMES = frozenset({
        'imagedb', 'colordb', 'fontdb', 'fontsizedb',
        'appearancedb', 'facetKeysdb', 'bitmapKeydb',
        'zcbezeldb', 'zcglyphdb'
    })

    def __init__(self, data: bytes | bytearray | memoryview):
        self._data = memoryview(data).cast("B")
        self.header = self._read_header()
        self.blocks = self._read_block_index()
        self.variables = self._read_variables()

    @classmethod
    def from_path(cls, path: Path | str) -> "BOMStore":
        return cls(Path(path).read_bytes())

    def _range(self, offset: int, length: int, label: str) -> memoryview:
        if offset < 0 or length < 0 or offset > len(self._data) or length > len(self._data) - offset:
            raise BOMError(f"{label} range is outside the file: offset={offset}, length={length}")
        return self._data[offset:offset + length]

    def _read_header(self) -> BOMHeader:
        if len(self._data) < self.HEADER.size:
            raise BOMError("file is shorter than the 32-byte BOM header")
        magic, version, count, index_off, index_len, vars_off, vars_len = self.HEADER.unpack_from(self._data)
        if magic != self.MAGIC:
            raise BOMError(f"invalid BOM magic: {bytes(magic)!r}")
        if version != 1:
            raise BOMError(f"unsupported BOMStore version: {version}")
        self._range(index_off, index_len, "block index")
        self._range(vars_off, vars_len, "variables")
        return BOMHeader(version, count, index_off, index_len, vars_off, vars_len)

    def _read_block_index(self) -> dict[int, Block]:
        raw = self._range(self.header.index_offset, self.header.index_length, "block index")
        if len(raw) < 4:
            raise BOMError("block index is truncated")
        capacity = struct.unpack_from(">I", raw, 0)[0]
        # Index layout starts with a capacity word, followed by `capacity`
        # (offset,length) pairs. Identifier zero is the null block. A free-list
        # trailer may follow the allocated pair array.
        required = 4 + capacity * 8
        if capacity > 1_000_000 or required > len(raw):
            raise BOMError(f"invalid block index capacity: {capacity}")
        result: dict[int, Block] = {}
        for identifier in range(1, capacity):
            offset, length = struct.unpack_from(">II", raw, 4 + identifier * 8)
            if offset == 0 and length == 0:
                continue
            self._range(offset, length, f"block {identifier}")
            result[identifier] = Block(identifier, offset, length)
        return result

    def _read_variables(self) -> dict[str, int]:
        raw = self._range(self.header.variables_offset, self.header.variables_length, "variables")
        if len(raw) < 4:
            raise BOMError("variables table is truncated")
        count = struct.unpack_from(">I", raw, 0)[0]
        cursor = 4
        result: dict[str, int] = {}
        if count > 100_000:
            raise BOMError(f"unreasonable variable count: {count}")
        for _ in range(count):
            if cursor + 5 > len(raw):
                raise BOMError("variables table entry is truncated")
            identifier = struct.unpack_from(">I", raw, cursor)[0]
            name_length = raw[cursor + 4]
            cursor += 5
            if cursor + name_length > len(raw):
                raise BOMError("variable name is truncated")
            name_bytes = bytes(raw[cursor:cursor + name_length])
            cursor += name_length
            try:
                name = name_bytes.decode("utf-8")
            except UnicodeDecodeError as exc:
                raise BOMError("variable name is not UTF-8") from exc
            if identifier not in self.blocks:
                raise BOMError(f"variable {name!r} references missing block {identifier}")
            if name in result:
                raise BOMError(f"duplicate variable name: {name!r}")
            result[name] = identifier
        return result

    def block(self, identifier: int) -> memoryview:
        try:
            item = self.blocks[identifier]
        except KeyError as exc:
            raise BOMError(f"unknown block identifier: {identifier}") from exc
        return self._range(item.offset, item.length, f"block {identifier}")

    def named_block(self, name: str) -> memoryview:
        try:
            identifier = self.variables[name]
        except KeyError as exc:
            raise BOMError(f"unknown named block: {name}") from exc
        return self.block(identifier)

    def get_databases(self) -> dict[str, int]:
        """Return a mapping of known CoreUI database names to their block identifiers.

        CoreUI uses multiple specialized databases for different types of assets.
        This method identifies which databases are present in this BOMStore.
        """
        return {
            name: identifier
            for name, identifier in self.variables.items()
            if name in self.DATABASE_NAMES
        }

    def has_database(self, name: str) -> bool:
        """Check if a specific CoreUI database is present."""
        return name in self.variables and name in self.DATABASE_NAMES

File: actool_rs/src/bom.rs
use byteorder::{BigEndian, ByteOrder};
use std::collections::HashMap;
use std::path::Path;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum BOMError {
    #[error("File is shorter than 32-byte BOM header")]
    TruncatedHeader,
    #[error("Invalid BOM magic: {0:?}")]
    InvalidMagic(Vec<u8>),
    #[error("Unsupported BOMStore version: {0}")]
    UnsupportedVersion(u32),
    #[error("{label} range outside file: offset={offset}, length={length}")]
    OutOfRange {
        label: &'static str,
        offset: usize,
        length: usize,
    },
    #[error("Block index capacity invalid or truncated: {0}")]
    InvalidCapacity(u32),
    #[error("Variables table entry truncated or count too high: {0}")]
    InvalidVariables(u32),
    #[error("Unknown block identifier: {0}")]
    UnknownBlock(u32),
    #[error("Unknown named block: {0}")]
    UnknownName(String),
    #[error("Variable name is not valid UTF-8")]
    InvalidUtf8,
    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BOMHeader {
    pub version: u32,
    pub block_count_hint: u32,
    pub index_offset: usize,
    pub index_length: usize,
    pub variables_offset: usize,
    pub variables_length: usize,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Block {
    pub identifier: u32,
    pub offset: usize,
    pub length: usize,
}

pub struct BOMStore {
    data: Vec<u8>,
    pub header: BOMHeader,
    pub blocks: HashMap<u32, Block>,
    pub variables: HashMap<String, u32>,
}

impl BOMStore {
    pub const MAGIC: &'static [u8; 8] = b"BOMStore";

    pub fn from_bytes(data: Vec<u8>) -> Result<Self, BOMError> {
        if data.len() < 32 {
            return Err(BOMError::TruncatedHeader);
        }

        if &data[0..8] != Self::MAGIC {
            return Err(BOMError::InvalidMagic(data[0..8].to_vec()));
        }

        let version = BigEndian::read_u32(&data[8..12]);
        if version != 1 {
            return Err(BOMError::UnsupportedVersion(version));
        }

        let block_count_hint = BigEndian::read_u32(&data[12..16]);
        let index_offset = BigEndian::read_u32(&data[16..20]) as usize;
        let index_length = BigEndian::read_u32(&data[20..24]) as usize;
        let variables_offset = BigEndian::read_u32(&data[24..28]) as usize;
        let variables_length = BigEndian::read_u32(&data[28..32]) as usize;

        let check_range = |offset: usize, length: usize, label: &'static str| {
            if offset > data.len() || length > data.len().saturating_sub(offset) {
                Err(BOMError::OutOfRange {
                    label,
                    offset,
                    length,
                })
            } else {
                Ok(())
            }
        };

        check_range(index_offset, index_length, "block index")?;
        check_range(variables_offset, variables_length, "variables")?;

        let header = BOMHeader {
            version,
            block_count_hint,
            index_offset,
            index_length,
            variables_offset,
            variables_length,
        };

        // Parse block index
        let raw_index = &data[index_offset..index_offset + index_length];
        if raw_index.len() < 4 {
            return Err(BOMError::InvalidCapacity(0));
        }
        let capacity = BigEndian::read_u32(&raw_index[0..4]);
        let required = 4 + (capacity as usize) * 8;
        if capacity > 1_000_000 || required > raw_index.len() {
            return Err(BOMError::InvalidCapacity(capacity));
        }

        let mut blocks = HashMap::new();
        for identifier in 1..capacity {
            let entry_off = 4 + (identifier as usize) * 8;
            let offset = BigEndian::read_u32(&raw_index[entry_off..entry_off + 4]) as usize;
            let length = BigEndian::read_u32(&raw_index[entry_off + 4..entry_off + 8]) as usize;
            if offset == 0 && length == 0 {
                continue;
            }
            check_range(offset, length, "block payload")?;
            blocks.insert(
                identifier,
                Block {
                    identifier,
                    offset,
                    length,
                },
            );
        }

        // Parse variables
        let raw_vars = &data[variables_offset..variables_offset + variables_length];
        if raw_vars.len() < 4 {
            return Err(BOMError::InvalidVariables(0));
        }
        let count = BigEndian::read_u32(&raw_vars[0..4]);
        if count > 100_000 {
            return Err(BOMError::InvalidVariables(count));
        }

        let mut cursor = 4;
        let mut variables = HashMap::new();
        for _ in 0..count {
            if cursor + 5 > raw_vars.len() {
                return Err(BOMError::InvalidVariables(count));
            }
            let identifier = BigEndian::read_u32(&raw_vars[cursor..cursor + 4]);
            let name_len = raw_vars[cursor + 4] as usize;
            cursor += 5;
            if cursor + name_len > raw_vars.len() {
                return Err(BOMError::InvalidVariables(count));
            }
            let name_bytes = &raw_vars[cursor..cursor + name_len];
            cursor += name_len;

            let name = std::str::from_utf8(name_bytes)
                .map_err(|_| BOMError::InvalidUtf8)?
                .to_string();
            variables.insert(name, identifier);
        }

        Ok(Self {
            data,
            header,
            blocks,
            variables,
        })
    }

    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<Self, BOMError> {
        let bytes = std::fs::read(path)?;
        Self::from_bytes(bytes)
    }

    pub fn block_data(&self, identifier: u32) -> Result<&[u8], BOMError> {
        let blk = self
            .blocks
            .get(&identifier)
            .ok_or(BOMError::UnknownBlock(identifier))?;
        Ok(&self.data[blk.offset..blk.offset + blk.length])
    }

    pub fn named_block_data(&self, name: &str) -> Result<&[u8], BOMError> {
        let id = self
            .variables
            .get(name)
            .ok_or_else(|| BOMError::UnknownName(name.to_string()))?;
        self.block_data(*id)
    }
}
